Update query building for Rails 6.1 compatibility

zokioki opened the pull request:
As of rails 6.1, the `where_clause_factory` method has been [removed](https://github.com/rails/rails/commit/550ce69f8adcd19542453ae03ac837a9ac8adf4b)
in favor of `build_where_clause` - this updates the logic to reflect it.

Addresses https://github.com/palkan/pgrel/issues/13.

zokioki committed:
update query building for rails 6.1 compatibility

As of rails 6.1, the `where_clause_factory` method has been removed
in favor of `build_where_clause` - this updates the logic to reflect it.

Addresses https://github.com/palkan/pgrel/issues/13

zokioki committed:
allow build_where_clause access to descendant classes

zokioki committed:
move build_where_clause to ancestor

palkan commented:
Thanks!
Let me fix the CI issues, and then we'll be able to merge this.

palkan commented:
OK, looks like CI is fixed.

Could you please rebase your branch on top of the new master?

zokioki commented:
@palkan Rebased changes - looks like the tests picked up another Rails 6 related failure regarding `Arel::Nodes::Equality#children` - will take a look at this.

zokioki committed:
update predicate parsing to reflect rails 6.1 updates

The `WhereClause#ast` method has been changed as of Rails 6.1 to
return only the first predicate if only a single one is present [1].
This means that the returned value from `ast` will not always be
an instance of `Arel::Nodes::And` and we can't therefore always
count on having `children` to iterate on.

This updates the logic to handle the case where a non-And type of node
is returned by `ast`.

[1] https://github.com/rails/rails/pull/39046/commits/699b64ab6322a549a84c7d116fe18cbb522ce64f

zokioki committed:
convert HomogenousIn node type to In to retain left assignment

A new `Arel::Nodes::HomogeneousIn` node type was introduced in Rails 6.1 [1],
which breaks the logic in `where_with_prefix` as this new type does not allow
for setting its `left` attribute, among other things.

This change converts the Homogeneous node to its `Arel::Nodes::In` counterpart,
which allows the existing logic to assign the sql literal as before. This does
feel like a bit of a quick fix and a refactor may be good at some point to
more soundly accommodate the newly introduced node type.

[1] https://github.com/rails/rails/commit/72fd0bae5948c1169411941aeea6fef4c58f34a9

zokioki committed:
only check for HomogeneousIn type if defined

zokioki commented:
@palkan Updated to accommodate the [newly added](https://github.com/rails/rails/commit/72fd0bae5948c1169411941aeea6fef4c58f34a9) `Arel::Nodes::HomogeneousIn` node type, CI is now passing.

palkan commented:
Thanks!
